regex-syntax/src/parser.rs

```diff
@@ -596,27 +663,50 @@ impl Parser {
                 }
             }
         }
-        class = self.class_transform(negated, class).canonicalize();
-        if class.is_empty() {
-            // e.g., [^\d\D]
-            return Err(self.err(ErrorKind::EmptyClass));
-        }
-        Ok(Build::Expr(if self.flags.unicode {
-            Expr::Class(class)
+        Ok(if self.flags.casei {
+            // Case folding canonicalizes too
+            class.case_fold()
         } else {
-            let byte_class = class.to_byte_class();
+            class.canonicalize()
+        })
+    }
 
-            // If `class` was only non-empty due to multibyte characters, the
-            // corresponding byte class will now be empty.
-            //
-            // See https://github.com/rust-lang/regex/issues/303
-            if byte_class.is_empty() {
-                // e.g., (?-u)[^\x00-\xFF]
-                return Err(self.err(ErrorKind::EmptyClass));
+    // Parses an escape in a character class.
+    //
+    // This is a helper for `parse_class`. Instead of returning an `Ok` value,
+    // it either mutates the char class or returns an error.
+    //
+    // e.g., `\wx`
+    //
+    // Start: `\`
+    // End:   `x`
+    fn parse_class_escape(&mut self, class: &mut CharClass) -> Result<()> {
+        match try!(self.parse_escape()) {
+            Build::Expr(Expr::Class(class2)) => {
+                class.ranges.extend(class2);
             }
-
-            Expr::ClassBytes(byte_class)
-        }))
+            Build::Expr(Expr::ClassBytes(class2)) => {
```

Comment:
What are all these extra cases for? I don't see how they relate to the rest of the changes in this PR. Do they add support for more types of escapes in character classes? Which tests test these new branches?